pandas/_libs/tslibs/src/datetime/np_datetime_strings.c

```diff
@@ -67,49 +67,59 @@ This file implements string parsing and creation for NumPy datetime.
  * Returns 0 on success, -1 on failure.
  */
 
+typedef enum {
+    COMPARISON_SUCCESS,
+    COMPLETED_PARTIAL_MATCH,
+    COMPARISON_ERROR
+} DatetimePartParseResult;
 // This function will advance the pointer on format
 // and decrement characters_remaining by n on success
-// On failure will return -1 without incrementing
-static int compare_format(const char **format, int *characters_remaining,
-                 const char *compare_to, int n, const int exact) {
+// On failure will return COMPARISON_ERROR without incrementing
+// If `format_requirement` is PARTIAL_MATCH, and the `format` string has
+// been exhausted, then return COMPLETED_PARTIAL_MATCH.
+static DatetimePartParseResult compare_format(
+        const char **format,
+        int *characters_remaining,
+        const char *compare_to,
+        int n,
+        const FormatRequirement format_requirement
+) {
+  if (format_requirement == PARTIAL_MATCH && !*characters_remaining) {
```

Comment:
also a nit but I think we need to handle characters_remaining being 0. It could just simply return a `COMPARISON_ERROR` right?

Understood it is impossible in the current state of things. However, if this gets refactored in the future and a negative number makes its way in here uncaught I _think_ it would return a `COMPARISON_SUCCCESS` and be very difficult to troubleshoot without intimate knowledge of this function

Reply:
yup, thanks for this (and other) thoughtful comments!

Reply:
> also a nit but I think we need to handle characters_remaining being 0

I presume you meant "less than 0" - that's what I've gone for anyway